Docs: Rewrite the "Row sorting" guide

kirszenbaum opened the pull request:
This PR:
- Rewrites the "Row sorting" guide in line with the [GitBook draft](https://app.gitbook.com/o/-LAYxy5uw1mD-6fQSEXj/s/LTM2f9nnKNQWY3SGiyZU/rows-sorting) agreed between @krzysztofspilka and @kirszenbaum.

[skip changelog]

kirszenbaum commented:
@kirszenbaum test

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1,79 +1,141 @@
+description: Sort rows alphabetically or numerically, in ascending, descending or a custom order, across one or multiple columns.
 permalink: /row-sorting
 canonicalUrl: /row-sorting
+tags:
```
Cool, but I would add some of the following:

- row ordering
- order by
- sort ascending
- sort descending
- highest to lowest
- unordered data
- sortEmptyCells

I would remove
- columns sorting (repetition)

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1,79 +1,141 @@
+# Rows sorting
 
-Sort your data in ascending, descending, or a custom order, across one or multiple rows of the grid.
+<p style="font-size:20px;">Sort rows alphabetically or numerically, in ascending, descending or a custom order, across one or multiple columns.</p>
```
Let's remove the inline style and restore the "Overview" heading with a more extensive description, like so:

> ## Overview 
> Handsontable's sorting feature allows you to easily rearrange rows of data based on the values in specified columns. You can sort rows alphabetically, numerically, or based on the data type set in the configuration. You can choose to sort rows in ascending or descending order or define a custom order to suit your needs. This feature is especially useful for analyzing and organizing large datasets, helping you to identify patterns and trends.

The text above might be adjusted however you wish, just keep it more descriptive than it was.

kirszenbaum replied:
How about this? [c72bac8](https://github.com/handsontable/handsontable/pull/10183/commits/c72bac8f8c65bacf61c825482506545723214003)

https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/a/allow-allows
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/l/let-lets

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1,79 +1,141 @@
-* A sorting indicator will be enabled
-* Empty cells won't be sorted
-* The sort method will use default compare functions - read more about them [here](#default-compare-functions-sorting-different-kinds-of-data)
+Try it out: click on a column label to sort the rows in ascending, descending, or the original order.
```
The more accessible approach, I think:
> In the demo showcase below presents the single-column sorting mode. Click on a column name to sort the data in a column in ascending, descending, or restore the original order.

kirszenbaum replied:
1. If possible, I'd avoid expresssions with "below" or "above":
    - If the layout changes, we need to change the content
    - It sounds stiff (formal register)
    - The MS style guide advises against it [as well](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/b/below)
2. "single-column sorting mode" – we're not explaining this term, we're not using it anywhere else.

How about this? [c72bac8](https://github.com/handsontable/handsontable/pull/10183/commits/c72bac8f8c65bacf61c825482506545723214003)

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1,79 +1,141 @@
-* The sort method will use default compare functions - read more about them [here](#default-compare-functions-sorting-different-kinds-of-data)
+Try it out: click on a column label to sort the rows in ascending, descending, or the original order.
 
 ::: only-for javascript
```
Please remove the initial order from the column "Model" - it might be misleading.

kirszenbaum replied:
Done: [4b83829](https://github.com/handsontable/handsontable/pull/10183/commits/4b83829575eae775896a8915b357662ff2e343a4)

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -83,93 +145,502 @@ import 'handsontable/dist/handsontable.full.min.css';
+    columnSorting: true,
+
+    // exclude rows 1 and 2 from sorting
+    afterColumnSort() {
```
This is great. However, the typical usage is when someone wants to exclude all frozen rows from the sorted range. Can we create an additional example in which we show to exclude those - without passing an exact number of frozen rows?

kirszenbaum replied:
Done

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -83,93 +145,502 @@ import 'handsontable/dist/handsontable.full.min.css';
+
 :::
 
+Only the table view is sorted. Your source data remains in the original order.
```
What's "table view"? I think we're introducing a new nomenclature here for a thing that already has the name "visual data" or similar. To be determined.

kirszenbaum replied:
Done: [b545d9e](https://github.com/handsontable/handsontable/pull/10183/commits/b545d9e5f9253748351c0e786b8d3d45d4af535d)

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -297,153 +982,253 @@ import 'handsontable/dist/handsontable.full.min.css';
+:::
 
-## Custom compare functions
+## Set an initial sort order
```
We need to extend this section a bit to make it more meaningful and complete. We can do this by adding a showcase in which we present the multi-column sorting by two columns (Brand and Model) in ascending order (Brand), and descending order (Model).

Avoid using column indices. Use names instead.

kirszenbaum replied:
Ok, how about this: [661cc00](https://github.com/handsontable/handsontable/pull/10183/commits/661cc0036c10ebec5e8e2ea583c85762a6fe58df)

krzysztofspilka replied:
> Avoid using column indices. Use names instead.
Even in the comments section, I would write "sort column "Model" ascending" instead of "sort column 1", especially that "column 1" means "index 0" which might not be that obvious for some of the users.

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
````diff
@@ -453,163 +1238,919 @@ import 'handsontable/dist/handsontable.full.min.css';
+
+```css
+/* the icon for ascending order */
+.your-handsontable-instance span.colHeader.columnSorting.ascending::before {
````
`.your-handsontable-instance` doesn't make sense here. I can have a custom class, but that doesn't mean it should be included here.

Let's replace it with `.handsontable` class.

Also, there are two methods to replace that default icon. **The recommended way** should be to replace the default image (currently, it's a base64 image, but it could be plain PNG).

**The alternative way** should be to place there a custom emoji through the 'content` property in CSS.

In the alternative method, I would replace the 👆 icon with △ and 👇 with ▽.

Also, let's use the demo with refreshed data set in this section.

kirszenbaum replied:
Ok, done: [1201b6f](https://github.com/handsontable/handsontable/pull/10183/commits/1201b6fd6915c303dc49cb2cb9735a5ac21f8d5f)

Mind that when we change the `background-image` property, the icon displays too close to the column name

krzysztofspilka replied:
> Mind that when we change the background-image property, the icon displays too close to the column name

@kirszenbaum If the new icon has a size other than the original `height: 10px`; `width: 5px`, you need to adjust CSS - [view the demo](https://jsfiddle.net/70jLmxnc/25/).

Some arrows in base64 for the testing purposes are available here: https://iconmonstr.com/angel-up-circle-thin-svg/ (click embed > base64)

kirszenbaum replied:
Right, thanks! Fixed: [398b56a](https://github.com/handsontable/handsontable/pull/10183/commits/398b56acc314ca830651ff5dd90da5a08122444c)

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -453,163 +1238,919 @@ import 'handsontable/dist/handsontable.full.min.css';
+:::
 
-    // const newData = ... // Calculated data set, ie. from an AJAX call.
+## Control sorting programmatically
```
What developers want to do, programatically, with sorting:
- Enable / disable for the entire grid and particular columns
- Sort particular columns ascending or descending, or to restore the original state. That includes multi-column sorting.

kirszenbaum replied:
Ok, done

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -453,163 +1238,919 @@ import 'handsontable/dist/handsontable.full.min.css';
+## Related resources
 
-List of methods exposed by the plugin:
+Read our [blog article](https://handsontable.com/blog/articles/2018/11/feature-spotlight-multi-column-sorting) about multi-column rows sorting.
```
This blog post is outdated imho and should be removed from here. Please check that with the original author (@jansiegel) and @AMBudnik

kirszenbaum replied:
Ok, done

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -453,163 +1238,919 @@ import 'handsontable/dist/handsontable.full.min.css';
-* [`isSorted`](@/api/columnSorting.md#issorted)
-* [`setSortConfig`](@/api/columnSorting.md#setsortconfig)
-* [`sort`](@/api/columnSorting.md#sort)
+### More examples
```
We can add this section later on. We need to determine the default playground used here first.

krzysztofspilka reviewed (changes requested):
@kirszenbaum I've done the first review aka "first cats behind the fence" :D

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -83,93 +145,502 @@ import 'handsontable/dist/handsontable.full.min.css';
 
+Only the table view is sorted. Your source data remains in the original order.
+
+## Enable rows sorting
```
Done

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -83,93 +145,502 @@ import 'handsontable/dist/handsontable.full.min.css';
+        inStock: true,
+      },
+    ]}
+      columns={[
```
@krzysztofspilka – hm, what property name should I use for that? (`colHeaders` doesn't seem to work when used inside `columns`)

kirszenbaum replied:
Done, for all demos: [2e4418d](https://github.com/handsontable/handsontable/pull/10183/commits/2e4418dd88930a4fca1aa265143eda8ad9b3e5ce).

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -83,93 +145,502 @@ import 'handsontable/dist/handsontable.full.min.css';
+
+:::
+
+### Import the sorting module
```
As of now, none of the examples import the `ColumnSorting` module, so the whole **Import the sorting module** section is an alternative to what's shown in the examples.

Should I rework all the examples to use the modularized approach?

krzysztofspilka replied:
@kirszenbaum yep, you might be right. We can't just change that one page because that would ruin the consistency of the guides in this matter. I would do two things to emphasize that we have modules:

1. In the first demo on the page, I would put a commented-out line indicating that there is a way to import the module more precisely. Discuss this with the devs, please.

Now:
```
import Handsontable from 'handsontable';
import 'handsontable/dist/handsontable.full.min.css';

const container = document.querySelector('#example1');
```

Then
```
import Handsontable from 'handsontable/base';

// import all modules

// commented out below import all modules: import and register only the required modules used in the demo, such as cell type: text, date, number, time, checkbox; column sorting (or multi-column sorting depending on the demo)

```

2. The "Import (module name)" is a common section for all guides, or at least most of them. It could be placed below the core content and become a part of our template.

kirszenbaum replied:
That'd be quite a few lines to comment out. Here's what I believe needs to happen in each configuration:

## JS – full version

```js
import Handsontable from 'handsontable';

import 'handsontable/dist/handsontable.full.min.css';
```

## JS – through individual modules

Here, you need to import modules from `handsontable/base` instead of `handsontable`:

```js
import Handsontable from 'handsontable/base';

import 'handsontable/dist/handsontable.full.min.css';

import {
  registerPlugin,
  ColumnSorting,
  MultiColumnSorting,
} from 'handsontable/plugins';

registerPlugin(ColumnSorting);
registerPlugin(MultiColumnSorting);
```

## React – full version

Here, you need to register all modules anyway:

```jsx
import { HotTable } from '@handsontable/react';
import { registerAllModules } from 'handsontable/registry';
import 'handsontable/dist/handsontable.full.min.css';

registerAllModules();
```

## React – through individual modules

```jsx
import { HotTable } from '@handsontable/react';
import Handsontable from 'handsontable/base';

import 'handsontable/dist/handsontable.full.min.css';

import {
  registerPlugin,
  ColumnSorting,
  MultiColumnSorting,
} from 'handsontable/plugins';

registerPlugin(ColumnSorting);
registerPlugin(MultiColumnSorting);
```

How about this then:
1. We stick to one way throughout the page (and documentation): full, non-modularized version.
2. We add **Import the sorting module** at the end of the page, but we link to it from the **Enable the sorting module** section.
3. In the first demo, we add a single commented-out line: "`// to import just individual modules, see the 'Import the sorting module' section of this page`"

[0534ce9](https://github.com/handsontable/handsontable/pull/10183/commits/0534ce9b92e86272226d6b209ed92dedfe05fcd2)

krzysztofspilka replied:
> 1. We stick to one way throughout the page (and documentation): full, non-modularized version.

Switching to modules is probably a subject for another project. We can stick with the current approach.

> 2. We add Import the sorting module at the end of the page, but we link to it from the Enable the sorting module section.

Enabling takes place after initialization, so I would skip the link and place the "importing" section somewhere at end of the article.

> In the first demo, we add a single commented-out line: "// to import just individual modules, see the 'Import the sorting module' section of this page"

Good idea, a little remix though: "// to import sorting as individual module, see the 'Import the sorting module' section on this page"

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -83,93 +145,502 @@ import 'handsontable/dist/handsontable.full.min.css';
+    columnSorting: {
+      // let the end user sort rows by clicking on the column label
+      headerAction: true,
+      // don't sort empty cells (move rows with empty cells to the bottom)
```
Done

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -83,93 +145,502 @@ import 'handsontable/dist/handsontable.full.min.css';
+      headerAction: true,
+      // don't sort empty cells (move rows with empty cells to the bottom)
+      sortEmptyCells: false,
+      // show the sort order indicator in the column header
```
Done

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -83,93 +145,502 @@ import 'handsontable/dist/handsontable.full.min.css';
+To enable rows sorting, use the [`columnSorting`](@/api/options.md#columnsorting) option.
 
-## Extended plugin configuration
+::: only-for javascript
```
Ok, done: [8b3cc93](https://github.com/handsontable/handsontable/pull/10183/commits/8b3cc9319d2e8e07ce558221595615952bc854f6)

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -179,115 +650,329 @@ import 'handsontable/dist/handsontable.full.min.css';
+
+:::
+
+## Sort by multiple columns
```
Got it. How about this? [ab811d6](https://github.com/handsontable/handsontable/pull/10183/commits/ab811d673c8f669388c0949b8af7631efaef9c9c)

krzysztofspilka replied:
@kirszenbaum I will review all your changes at once if you don't mind. Please pass this PR to me after it's finished. Thanks!

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1160,12 +1131,19 @@ const yourHandsontableInstance = new Handsontable(container, {
       sortOrder: 'asc',
-    },
+      },
+      // at initialization, sort rows by column 2, in descending order
```
Better than `column 2` ⟶ column name

kirszenbaum replied:
Ok, done: [1e8d91d](https://github.com/handsontable/handsontable/pull/10183/commits/1e8d91d1e72237254a75bb2b33d30de604a7ddf0)

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -453,163 +1184,2251 @@ import 'handsontable/dist/handsontable.full.min.css';
+
+## Control sorting programmatically
+
+You can control sorting at runtime by using Handsontable's API.
```
...to do what? What's the benefit or a use case? We need a better context here.

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -453,163 +1184,2251 @@ import 'handsontable/dist/handsontable.full.min.css';
+To sort data programmatically, use the [`columnSorting.sort()`](@/api/columnSorting.md#sort) method.
+Remember to [enable sorting](#enable-sorting) first.
+
+Mind that using [`columnSorting.sort()`](@/api/columnSorting.md#sort) erases any previous sort orders.
```
erases? Maybe: overrides?

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -453,163 +1184,2251 @@ import 'handsontable/dist/handsontable.full.min.css';
+
+### Exclude rows from sorting
+
+You can exclude any number of top or bottom rows from sorting.
```
Let's give it a business context - you may want to do that in case of fixed rows at the top (usually used as column names then), or at the bottom (used for summary calculations). I would put the demo there with both frozen rows at the top and the bottom - must scroll!

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
````diff
@@ -453,163 +1184,2251 @@ import 'handsontable/dist/handsontable.full.min.css';
+```
+:::
+
+In this demo, click on any column name. The rows get sorted, but the the first row and the last row stay in place.
````
Please rewrite this sentence to avoid generic sentences like "In this demo, click on any column name.".

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -453,163 +1184,2251 @@ import 'handsontable/dist/handsontable.full.min.css';
+
+:::
+
+You can also exclude [frozen rows](@/guides/rows/row-freezing.md) from sorting. In the following demo,
```
> Instead of providing an exact number of rows excluded from sorting, you can exclude all frozen rows at once. This is helpful if end users can dynamically adjust the number of frozen rows in your application.

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -453,163 +1184,2251 @@ import 'handsontable/dist/handsontable.full.min.css';
+## Import the sorting module
 
-List of methods exposed by the plugin:
+If you're using Handsontable through [modules](@/guides/tools-and-building/modules.md),
```
> To use the sorting feature, you don't need to import all modules. You save on our bundle size by importing and registering only required modules, in this that would be `base` together with `ColumnSorting`, `MultiColumnSorting`, or both if you prefer.

krzysztofspilka commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -453,163 +1184,2251 @@ import 'handsontable/dist/handsontable.full.min.css';
-- Plugins:
-  - [`ColumnSorting`](@/api/columnSorting.md)
-  - [`MultiColumnSorting`](@/api/multiColumnSorting.md)
+- View [GitHub issues](https://github.com/handsontable/handsontable/labels/Column%20sorting) related to sorting.
```
1. It suggests that the sorting feature has "issues" (meaning bugs?). I would replace it with "View related topics on GitHub" or similar
2. I would remove the full-stop symbols from the end of sentences because they are, in fact, not sentences.
3. Link should be CTAs (call to action), so in a sentence "Report a new GitHub issue" the link should be applied to CTA words, e.g. [Report an issue]() on GitHub; [Ask a question]() on StackOverflow; [Start discussion]() on forum; [Contact our technical support]() to get help

krzysztofspilka reviewed (changes requested):
Almost there @kirszenbaum :) Once all comments are addressed and solved, please make someone from the dev team a final reviewer. Adam will help you with this. Thanks!

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1,176 +1,609 @@
 ## Overview
 
-The [`ColumnSorting`](@/api/columnSorting.md) plugin works as a middleware between the data source and the Handsontable rendering module.
+With sorting, you can easily rearrange rows of data, based on the values in specified columns.
```
"Selected columns" may sound more natural, but it's ambiguous. In the context of HoT, "selected columns" may mean [cell selection](https://handsontable.com/docs/react-data-grid/selection/). It could suggest that when you select a column (i.e., click on that column's header), the data gets sorted, which is not true (you need to click on the column's label/header).

I'd keep the word "selected" as a reserved, technical term, and use it consistently only in the context of cell/row/column selection per se.

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1,176 +1,609 @@
+### Demo
 
-## Basic plugin configuration
+Click on any column name. Based on the values in that column,
```
> Click on a column name in the column header 

(I think you meant "a column name in **a** column header) – "column name" is self-explanatory, no need to repeat that.

> to sort the values in the column

That's not true: you don't sort "the values in the column", you sort all the values in all the columns. It's the same problem as with calling the feature "column sorting" instead of "row sorting".

> **,** or go back to the original order.

That comma implies that you either "click" or "go back".

I'd go for:
> Click on a colum name to sort the values in ascending (↑) or descending (↓) order, or to go back to the original order.

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1,176 +1,609 @@
-const hot = new Handsontable(container, {
+
+const handsontableInstance = new Handsontable(container, {
   data: [
```
No. I created an issue for that: [#1125](https://github.com/handsontable/dev-handsontable/issues/1125).

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1,176 +1,609 @@
   ],
-  colHeaders: ['Brand', 'Model', 'Code name', 'Country of origin', 'Rank'],
-  columnSorting: true,
   height: 'auto',
```
`height: 'auto'` and ``width: 'auto'`` have been used across nearly all docs examples, and we didn't break with this practice so far, as we're not really sure how those options are supposed to work. See [#980](https://github.com/handsontable/dev-handsontable/issues/980) and [#1120](https://github.com/handsontable/dev-handsontable/issues/1120).

I'll remove this from the new "Rows sorting" page and we'll see if the examples still work correctly.

FYI @adamu-handsontable @Serrrek89

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1,176 +1,609 @@
       ]}
-      colHeaders={['Brand', 'Model', 'Code name', 'Country of origin', 'Rank']}
       columnSorting={true}
       height="auto"
```
See: https://github.com/handsontable/handsontable/pull/10183#discussion_r1108373959

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -1,176 +1,609 @@
+      className: 'htCenter',
+    },
+  ],
+  height: 'auto',
```
Removing – see https://github.com/handsontable/handsontable/pull/10183#discussion_r1108373959

kirszenbaum commented on docs/content/guides/rows/row-sorting.md:
```diff
@@ -297,154 +849,333 @@ import 'handsontable/dist/handsontable.full.min.css';
+You can set a default sort order that's applied every time Handsontable is launched.
 
-The plugin's options, such as `compareFunctionFactory`, `sortEmptyCells`, `headerAction`, `indicator`, can be set just for a particular column. This can be done by using the [`columns`](@/api/options.md#columns) option. The example below demonstrates how to disable the indicator and **completely block sorting action for the first column**:
+In this demo, the data is initially sorted:
```
As pointed out before, I wouldn't say "[below](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/b/below)", we can say "the following demo", though

aninde reviewed (changes requested):
I have ensured that the links and demos are functional in both the JS and React versions, but I have a few comments to make:

1. 9 out of the 10 demos within the code element in both JS and React versions overlaps another elements . They exceed the boundaries of the code element, causing overlap with the right navigation menu or adding a horizontal scrollbar on smaller screens.
The last one, "exclude-rows-from-sorting", appears to be correctly formatted within the code element in both JS and React versions. 

<img width="680" alt="Screenshot 2023-02-21 at 09 47 53" src="https://user-images.githubusercontent.com/30186818/220305342-c1009e75-cb4b-4935-bbd2-ac00c86f2c3f.png">


2. The multi-column sorting demo can only display two levels of multicolumn sorting due to the nature of the data. However, by making a small modification to the dataset, it can be made to show three or more levels. For instance, repeating the data in two columns for a single row would achieve this, as in the following example https://jsfiddle.net/aninde/jymgu3Lr/

<img width="350" alt="Screenshot 2023-02-21 at 09 59 58" src="https://user-images.githubusercontent.com/30186818/220305508-b8dc5c21-8d8d-4df4-93d6-26833cf7e532.png">

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3432 @@
+/* the icon for both ascending and descending order */
+.custom-sort-icon-1 span.colHeader.columnSorting.ascending::before,
+.custom-sort-icon-1 span.colHeader.columnSorting.descending::before {
+  background-image: url("../../../img/custom_sort_icon.png") !important;
```
```suggestion
  background-image: url("../../../img/custom_sort_icon.png") !important;
```
This demo does not work in jsfiddle or after copying/pasting the snippet to the developer's app. I'd suggest converting the icon file to base64 to get rid of mentioned problems.

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3432 @@
+
+:::
+
+## Add custom sort icons
```
WDYT adding a section about custom sort icons for multi-column sorting? Correct me if I'm wrong but right now I can see only a tutorial for single-column sorting.

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3432 @@
+    compareFunctionFactory: function(sortOrder, columnMeta) {
+      // implement your own comparator
+      return function(value, nextValue) {
+        if (`value` is less than `nextValue` by some ordering criterion) {
```
I don't think that it's a good idea to keep pseudo/meta code in code snippets. It's not copy-pastable, the color syntax - which is broken, reinforces the idea that something is wrong with that snippet.

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3432 @@
+
+const container = document.querySelector('#example8');
+
+const button_sort_ascending = document.querySelector('#sort_asc');
+const button_unsort = document.querySelector('#unsort');
+
+
+const handsontableInstance = new Handsontable(container, {
```
```suggestion
const buttonSortAscending = document.querySelector('#sort_asc');
const buttonUnsort = document.querySelector('#unsort');
const handsontableInstance = new Handsontable(container, {
```
I suggest using camelCase variable names for consistency with other examples within Docs.

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3432 @@
+
+const columnSortingPluginInstance = handsontableInstance.getPlugin('columnSorting');
+
+button_sort_ascending.addEventListener('click', () => {
```
```suggestion
buttonSortAscending.addEventListener('click', () => {
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3432 @@
+  });
+});
+
+button_unsort.addEventListener('click', () => {
```
```suggestion
buttonUnsort.addEventListener('click', () => {
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3432 @@
+import Handsontable from 'handsontable';
+import 'handsontable/dist/handsontable.full.min.css';
+
+const container = document.querySelector('#example9');
+
+const button_sort = document.querySelector('#sort');
+
+const handsontableInstance = new Handsontable(container, {
```
```suggestion
const container = document.querySelector('#example9');
const buttonSort = document.querySelector('#sort');
const handsontableInstance = new Handsontable(container, {
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3432 @@
+
+const multiColumnSortingPluginInstance = handsontableInstance.getPlugin('multiColumnSorting');
+
+button_sort.addEventListener('click', () => {
```
```suggestion
buttonSort.addEventListener('click', () => {
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3432 @@
+  colHeaders: true,
+  columnSorting: true,
+  // `afterColumnSort` is a Handsontable hook: it's fired after each sorting
+  afterColumnSort() {
+    // after each sorting, take row 1 and change its index to 0
+    handsontableInstance.rowIndexMapper.moveIndexes(
+      [
+        handsontableInstance.toVisualRow(0),
+      ], 0);
+
+    // after each sorting, take row 16 and change its index to 15
+    handsontableInstance.rowIndexMapper.moveIndexes(
+      [
+        handsontableInstance.toVisualRow(15),
+      ], 15);
+  },
+  cells(row, col, prop) {
+    if (row === 0) {
+      return {
+        type: 'text',
+        className: 'htCenter',
+      };
+    }
+    if (row === 15) {
+      return {
+        type: 'numeric',
+        className: 'htCenter',
+      };
+    }
+  },
```
```suggestion
  afterColumnSort() {
    const lastRowIndex = handsontableInstance.countRows() - 1;
    
    // after each sorting, take row 1 and change its index to 0
    handsontableInstance.rowIndexMapper
    	.moveIndexes(handsontableInstance
      	.toVisualRow(0), 0);

    // after each sorting, take row 16 and change its index to 15
    handsontableInstance.rowIndexMapper
    	.moveIndexes(handsontableInstance
      	.toVisualRow(lastRowIndex), lastRowIndex);
  },
  cells(row, col, prop) {
    const lastRowIndex = this.instance.countRows() - 1;
    
    if (row === 0) {
      return {
        type: 'text',
        className: 'htCenter',
      };
    }
    if (row === lastRowIndex) {
      return {
        type: 'numeric',
        className: 'htCenter',
      };
    }
  },
```
I changed the demo slightly, and I used API to freeze the last row based on the dataset length, not the hardcoded values. So it has a chance to work when the user starts changing the number of rows - for example.

budnix reviewed (changes requested):
I found some issues related to the code. All changes that I left in the code I'd apply to the React counterparts.

kirszenbaum commented:
@aninde, @budnix, thank you! I applied all the suggestions. Please mind there's a new demo in the **Add custom sort icons** section:
http://localhost:8080/docs/javascript-data-grid/rows-sorting/#add-custom-sort-icons
http://localhost:8080/docs/react-data-grid/rows-sorting/#add-custom-sort-icons

budnix reviewed (approved):
Great! 👌

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3903 @@
+          sellDate: '02/10/2023',
+          sellTime: '13:23',
+          inStock: true,
+        },
+    ]}
+      columns={[
```
```suggestion
        },
      ]}
      columns={[
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3903 @@
+          sellDate: '02/10/2023',
+          sellTime: '13:23',
+          inStock: true,
+        },
+    ]}
+      columns={[
```
```suggestion
        },
      ]}
      columns={[
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3903 @@
+          sellDate: '02/10/2023',
+          sellTime: '13:23',
+          inStock: true,
+        },
+    ]}
+      columns={[
```
```suggestion
        },
      ]}
      columns={[
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3903 @@
+          sellDate: '02/10/2023',
+          sellTime: '13:23',
+          inStock: true,
+        },
+    ]}
+      columns={[
```
```suggestion
        },
      ]}
      columns={[
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3903 @@
+          sellDate: '02/10/2023',
+          sellTime: '13:23',
+          inStock: true,
+        },
+    ]}
+      columns={[
```
```suggestion
        },
      ]}
      columns={[
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3903 @@
+          sellDate: '02/10/2023',
+          sellTime: '13:23',
+          inStock: true,
+        },
+    ]}
+      columns={[
```
```suggestion
        },
      ]}
      columns={[
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
```diff
@@ -0,0 +1,3903 @@
+export const HandsontableComponent = () => {
+  const hotTableComponentRef = useRef(null);
+  const exclude = () => {
+    // take row 1 and change its index to 0
+    hotTableComponentRef.current.hotInstance.rowIndexMapper.moveIndexes(
+      [
+        hotTableComponentRef.current.hotInstance.toVisualRow(0),
+      ], 0);
+
+    // take row 16 and change its index to 15
+    hotTableComponentRef.current.hotInstance.rowIndexMapper.moveIndexes(
+      [
+        hotTableComponentRef.current.hotInstance.toVisualRow(15),
+      ], 15);
```
```suggestion
    const handsontableInstance = hotTableComponentRef.current.hotInstance;
    const lastRowIndex = handsontableInstance.countRows() - 1;
    
    // after each sorting, take row 1 and change its index to 0
    handsontableInstance.rowIndexMapper
    	.moveIndexes(handsontableInstance
      	.toVisualRow(0), 0);
    // after each sorting, take row 16 and change its index to 15
    handsontableInstance.rowIndexMapper
    	.moveIndexes(handsontableInstance
      	.toVisualRow(lastRowIndex), lastRowIndex);
```

budnix commented on docs/content/guides/rows/rows-sorting.md:
````diff
@@ -0,0 +1,3903 @@
+```css
+/* the icon for both ascending and descending order */
+.custom-sort-icon-example-1 span.colHeader.columnSorting.ascending::before,
+.custom-sort-icon-example-1 span.colHeader.columnSorting.descending::before {
````
I looked deeper into that example, and I think found some optimizations worth apply.
```css
/* the icon for both ascending and descending order */
.custom-sort-icon-example-1 span.colHeader.columnSorting.ascending::before,
.custom-sort-icon-example-1 span.colHeader.columnSorting.descending::before {
  background-image: url(data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iaXNvLTg4NTktMSI/Pg0KPCEtLSBVcGxvYWRlZCB0bzogU1ZHIFJlcG8sIHd3dy5zdmdyZXBvLmNvbSwgR2VuZXJhdG9yOiBTVkcgUmVwbyBNaXhlciBUb29scyAtLT4NCjxzdmcgZmlsbD0iIzAwMDAwMCIgaGVpZ2h0PSI4MDBweCIgd2lkdGg9IjgwMHB4IiB2ZXJzaW9uPSIxLjEiIGlkPSJMYXllcl8xIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiANCgkgdmlld0JveD0iMCAwIDMzMCAzMzAiIHhtbDpzcGFjZT0icHJlc2VydmUiPg0KPHBhdGggaWQ9IlhNTElEXzIyNV8iIGQ9Ik0zMjUuNjA3LDc5LjM5M2MtNS44NTctNS44NTctMTUuMzU1LTUuODU4LTIxLjIxMywwLjAwMWwtMTM5LjM5LDEzOS4zOTNMMjUuNjA3LDc5LjM5Mw0KCWMtNS44NTctNS44NTctMTUuMzU1LTUuODU4LTIxLjIxMywwLjAwMWMtNS44NTgsNS44NTgtNS44NTgsMTUuMzU1LDAsMjEuMjEzbDE1MC4wMDQsMTUwYzIuODEzLDIuODEzLDYuNjI4LDQuMzkzLDEwLjYwNiw0LjM5Mw0KCXM3Ljc5NC0xLjU4MSwxMC42MDYtNC4zOTRsMTQ5Ljk5Ni0xNTBDMzMxLjQ2NSw5NC43NDksMzMxLjQ2NSw4NS4yNTEsMzI1LjYwNyw3OS4zOTN6Ii8+DQo8L3N2Zz4=) !important;
  /* minor adjustments, as the custom icon is of a different size than the original */
  top: 12px;
  right: -35px;
  width: 22px;
  height: 22px;
  zoom: .4;
}

/* the same icon as for ascending order, but rotated 180 degrees */
.custom-sort-icon-example-1 span.colHeader.columnSorting.descending:before {
  transform: scaleY(-1);
}
```
The changes contain:
 * Replace PNG with SVG - less code to present :)
 * Replace transform rotate with scaleY. With rotation, there are some glitches with descending sort order.

The same I'd apply to other demos that uses custom icons.

AMBudnik reviewed (approved):
I did not find anything that would force me to request for changes before publishing the page. But I have some suggestions

* Adding the list of cell types and the same demo as above seems to be redundant.
* I am not sure if we should mention the `Add custom sort icons` - it seems like something we cannot guarantee it won’t be changed in the future
* In the `Sort data programmatically` we mention about the ` columnSorting`, but in the code below, we define `columnSortingPluginInstance`. The same goes for the `Sort data programmatically by multiple columns` paragraph.
* We mention hooks in the API Reference section at the bottom but there’s not a single line of introduction to sorting hooks in the article.

kirszenbaum commented:
@AMBudnik – thanks for the feedback! As discussed on Slack:

> I am not sure if we should mention the Add custom sort icons - it seems like something we cannot guarantee it won’t be changed in the future

We're keeping the demo

> In the Sort data programmatically we mention about the  columnSorting, but in the code below, we define columnSortingPluginInstance. The same goes for the Sort data programmatically by multiple columns paragraph.

In the codes samples, I changed `columnSortingPluginInstance` to `columnSorting` and `multiColumnSortingPluginInstance` to `multiColumnSorting`

> We mention hooks in the API Reference section at the bottom but there’s not a single line of introduction to sorting hooks in the article.

I added a short new section about that: [Use sorting hooks](http://localhost:8080/docs/react-data-grid/rows-sorting/#use-sorting-hooks)

> Adding the list of cell types and the same demo as above seems to be redundant.

I'm updating the demo to feature all the cell types:
http://localhost:8080/docs/javascript-data-grid/rows-sorting/#sort-different-types-of-data
http://localhost:8080/docs/react-data-grid/rows-sorting/#sort-different-types-of-data

I've got one problem with the new demo, though: when I set `height: 'auto'`(which I need to keep the demo inside of the container), the behavior of the `handsontable` cells changes.

To illustrate this: In [this demo](https://jsfiddle.net/nzhug36d/), the `handsontable` cells properly display outside the container:
![1](https://user-images.githubusercontent.com/29247866/221551269-4afc3d0b-71e8-4bd6-a8a5-69b0dba69a7a.gif)

But when I set `height: 'auto'`, the `handsontable` cells get trimmed:
![2](https://user-images.githubusercontent.com/29247866/221551295-28ed5fdf-71f7-4bdb-bd61-4463095b8200.gif)

Styling the `handsontable` cell with `overflow: visible !important;` doesn't help: https://jsfiddle.net/emqufjrt/

How can I can keep `height: 'auto'`, but display `handsontable` cells in full? (perhaps @adamu-handsontable can help here)

aninde commented:
@kirszenbaum issue with displaying with height:'auto' is a known action reported as a bug https://github.com/handsontable/dev-handsontable/issues/439. We are at the stage of setting the specifications for this configuration https://github.com/handsontable/dev-handsontable/issues/1120.

aninde reviewed (changes requested):
One last thing to fix before publishing.
I think that link to  <a href="https://github.com/handsontable/handsontable/edit/develop/docs/content/guides/rows/rows-sorting.md" target="_blank" rel ="noopener noreferrer">Suggest edits</a>
on /docs/javascript-data-grid/rows-sorting/ and /docs/react-data-grid/rows-sorting/  should also contain `rel=nofollow`.

adamu-handsontable reviewed (approved):
Great work, thank you

kirszenbaum commented:
> One last thing to fix before publishing. I think that link to [Suggest edits](https://github.com/handsontable/handsontable/edit/develop/docs/content/guides/rows/rows-sorting.md) on /docs/javascript-data-grid/rows-sorting/ and /docs/react-data-grid/rows-sorting/ should also contain `rel=nofollow`.

@aninde – good idea. As discussed in the meeting, this requires further VuePress customization and affects all docs, so I'm opening a separate issue for this: [#1202](https://github.com/handsontable/dev-handsontable/issues/1202).
